### PR TITLE
[jmx] fix bugs when using AD and/or multiple config files

### PR DESCRIPTION
**Pull request by arbll**
### What does this PR do?

Fixes multiple bugs with jmx : 
- If multiple config files are present for the same integration, only one will be evaluated.
- When using the auto-discovery mechanism :
  + If multiple configs with the same name are discovered (two tomcat containers for ex.), only one will be evaluated.
  + If any auto-discovered config is unscheduled (when a container dies for ex.), jmxfetch will get killed.

The first bug comes from this part of the code : 
https://github.com/DataDog/datadog-agent/blob/188e0d024e2bdcb7ff8e8323a60b2bfd494e370c/pkg/collector/corechecks/embed/jmxloader.go#L34
where we save configs in a map by their name. This name is not unique and configs with the same name are getting overwritten.  

For the AD bugs, the problems are coming from the current architecture of the jmx checks lifecycle in the agent. 
Right now, we represent all the jmx checks as a [single `check.Check`](https://github.com/DataDog/datadog-agent/blob/master/pkg/collector/corechecks/embed/jmx.go) that is responsible for managing the jmxfetch process. 
This does not play well with the current implementation of the AD that assumes a `check.Check` structure represents a real check that can be stopped and not an aggregation of them. This is what causes the second AD bug described above.

This PR fixes these bugs by
1. Refactoring the jmx check lifecycle to have one `check.Check` by instance (It makes a copy of the config for each instance to allow them to be addressed individually)
2. Using the config hash/digest instead of the config name as an identifier.

### Testing

+ A docker image is available at `datadog/agent-dev:arbll-jmx-refactor-check-lifcycle-jmx`. (this image is based on `master` and will likely contain bugs, it should not be used in production)
+ To easily test jmx in AD scenarios I made a minimal tomcat+jmx docker image at `arbll/tomcat:latest`. It should work out of the box with the above image.

### CODE REVIEW

**Comment by olivielpeau** [comment]
@arbll I'm finally planning to review this, sorry for the delay... Before I do, could you rebase on top of master and add a release note?

File: pkg/collector/corechecks/embed/jmx/state.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+var state jmxState = jmxState{
+	configs:     cache.NewBasicCache(),
+	runnerError: make(chan struct{}),
+	runner:      &runner{},
+	lock:        &sync.Mutex{},
+}
+
+func (s *jmxState) scheduleCheck(c *JMXCheck) error {
+	s.lock.Lock()
```
Comment (truthbk) [inline pkg/collector/corechecks/embed/jmx/state.go]: `defer s.lock.Unlock()` would work well with this logic. Also, we could end up with a deadlock situation if return `err`. It's generally regarded as best practice unless you need to unlock in the function.

File: pkg/collector/corechecks/embed/jmx/state.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		}
+	}
+	s.configs.Add(string(c.id), c.config)
+
+	s.lock.Unlock()
+	return nil
+}
+
+func (s *jmxState) unscheduleCheck(c *JMXCheck) {
+	s.lock.Lock()
```
Comment (truthbk) [inline pkg/collector/corechecks/embed/jmx/state.go]: ditto regarding `defer s.lock.Unlock()`

File: pkg/collector/corechecks/embed/jmx/state.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func AddScheduledConfig(c integration.Config) {
+	state.configs.Add(fmt.Sprintf("%v_%v", c.Name, c.Digest()), c)
+}
+
+func GetScheduledConfigs() map[string]integration.Config {
+	configs := make(map[string]integration.Config, len(state.configs.Items()))
+	for name, config := range state.configs.Items() {
+		configs[name] = config.(integration.Config)
```
Comment (truthbk) [inline pkg/collector/corechecks/embed/jmx/state.go]: It's probably safe to assume the type cast will never fail....

File: pkg/collector/corechecks/embed/jmx/state.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+type jmxState struct {
+	configs     *cache.BasicCache
+	runnerError chan struct{}
+	runner      *runner
+	lock        *sync.Mutex
+}
+
+var state jmxState = jmxState{
```
[the remark this answers is not in the document]
Comment (truthbk) [reply pkg/collector/corechecks/embed/jmx/state.go]: Note: if you do refactor `jmxState` into `jmx.runner` (and remove `runner * runner`) from here `cache.BasicCache` and channels are both thread safe so you might be able to remove the lock.

File: pkg/collector/corechecks/embed/jmx/loader.go [hunk 5]
```diff
@@ -61,17 +76,18 @@ func (jl *JMXCheckLoader) Load(config integration.Config) ([]check.Check, error)
 	}
 
 	for _, instance := range config.Instances {
-		if err = jmxLauncher.Configure(instance, config.InitConfig); err != nil {
+		if err = state.runner.configureRunner(instance, config.InitConfig); err != nil {
```
Comment (truthbk) [inline pkg/collector/corechecks/embed/jmx/loader.go]: Now that we can know the state.... Does it make sense to try to configure the runner if the process is already running?

File: pkg/collector/corechecks/embed/jmx/loader_test.go [hunk 6]
```diff
@@ -58,18 +58,23 @@ func TestLoadCheckConfig(t *testing.T) {
...
+			checks = append(checks, c)
+		}
 	}
 
 	for _, cfg := range cfgs {
 		found := false
-		for k := range jmxLauncher.checks {
-			if k == fmt.Sprintf("%s.yaml", cfg.Name) {
+		for _, c := range checks {
+			if c.String() == cfg.Name {
```
Comment (truthbk) [inline pkg/collector/corechecks/embed/jmx/loader_test.go]: 🍰

**Review by truthbk — COMMENTED** [review]
Mostly looks good, I've added a few comments, some ideas regarding maybe a bit of a cleaner refactor where we'd encapsulate `state`. 

This looks like it will work for the most part. Nice one.

File: releasenotes/notes/fix-jmx-autodiscovery-1138ee8b3c7a8d20.yaml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+---
+
+fixes:
+  - Multiple config files can now be used for the same integration
+  - Multiple configs with the same name can now be resolved by the AD
+  - JMX is not killed when an AD config is unscheduled anymore
```
[the remark this answers is not in the document]
Comment (olivielpeau) [reply releasenotes/notes/fix-jmx-autodiscovery-1138ee8b3c7a8d20.yaml]: also, feel free to add more details on the last item, we could maybe say that `The JMXFetch process is now managed properly when JMXFetch configs are unscheduled through autodiscovery`

[the remark this answers is not in the document]
Comment (olivielpeau) [reply releasenotes/notes/fix-jmx-autodiscovery-1138ee8b3c7a8d20.yaml]: finally, I believe `AD` is an internal acronym, please write `autodiscovery` in release notes ;)

**Review by olivielpeau — COMMENTED** [review]
Made a few comments, overall looks great!

➕ on the architectural comment Jaime made. Could be nice to include in the PR depending on the bandwidth you have before the 6.3 freeze

**Comment by olivielpeau** [comment]
One more thing I forgot to mention earlier: currently the `status` command of the Agent displays under `Running Checks` stats for the de-scheduled JMXFetch checks that ran successfully, which is useless and could be pretty confusing for users.

One (quick and dirty) way to avoid that would be to not update the expvar stats (in the check runner) when the `Run` of one-time checks exits successfully. WDYT?

**Comment by arbll** [comment]
@olivielpeau Doing this for long-running checks (interval equal to 0) sounds OK to me as a quick fix. I feel like we are creating a lot of debt by forcefully making everything fit our check abstractions but it's probably not the moment to break everything yet.

**Comment by olivielpeau** [comment]
Agreed, we've extended the collector check abstraction more than the design really allows, but there's no other quick solution for now. Let's revisit this once the ongoing autodiscovery refactor is done.

File: pkg/collector/corechecks/embed/jmx/state.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	return nil
+}
+
+func (s *jmxState) unscheduleCheck(c *JMXCheck) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.configs.Remove(string(c.id))
+}
+
+func AddScheduledConfig(c integration.Config) {
```
Comment (truthbk) [inline pkg/collector/corechecks/embed/jmx/state.go]: You're still missing the comments for these exported functions, could you add them for `AddScheduledConfig` and `GetScheduledConfig` with a brief description?

**Review by truthbk — COMMENTED** [review]
Looks like we're mostly good to go. Could you please add a couple of comments for the exported functions? 

Also, let's try not to forget about `state` and eventually bring it into the runner, I feel that's its natural place. I know you don't have the bandwidth right now, but hopefully we can find the time in the next few weeks.

File: cmd/agent/app/jmx.go [hunk 9]
```diff
@@ -139,7 +142,13 @@ func setupAgent() error {
 }
 
 func runJmxCommand(command string) error {
-	err := setupAgent()
+	err := config.SetupLogger(jmxLogLevel, "", "", false, false, "", true, false)
```
Comment (truthbk) [inline cmd/agent/app/jmx.go]: :cake:

**Review by truthbk — APPROVED** [review]
Looks good to me! ✅

**Comment by truthbk** [comment]
Stealing your thunder @arbll! Shipping it :shipit:

**Pull request merged**